### PR TITLE
fix: disable ProvidePlugin for javascript/esm (fixes #7032)

### LINKED ISSUE #7032: ProvidePlugin injection into .mjs files seems to be broken
**Do you want to request a *feature* or report a *bug*?**

A bug, I think.

**What is the current behavior?**

At runtime, a `.mjs` module with injections from `ProvidePlugin` will throw an error at runtime, indicating that `require` is not defined. The final module code produced by webpack looks like this:

```javascript
/*! no static exports found */
/***/ (function(module, exports) {

/* WEBPACK VAR INJECTION */(function(foo) {console.log(foo);
/* WEBPACK VAR INJECTION */}.call(this, require("/Users/aparker/src/notes/webpack4upgrade/mjs-provide/foo.js")))

/***/ })

/******/ });
```

(Note the call to `require`.)

**If the current behavior is a bug, please provide the steps to reproduce.**

[You can reproduce the behavior using this repository.](https://github.com/xandris/webpack-provideplugin-mjs-bug)

**What is the expected behavior?**

The module should look more like this:

```javascript
/*! no static exports found */
/***/ (function(module, exports, __webpack_require__) {

/* WEBPACK VAR INJECTION */(function(foo) {console.log(foo);
/* WEBPACK VAR INJECTION */}.call(this, __webpack_require__(/*! ./foo.js */ "./foo.js")))

/***/ })

/******/ });
```

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

* `node -v`: `v9.11.1`
* `webpack -v`: `4.5.0`
* OS: Mac OS 10.13.4

### PR DESCRIPTION
**Pull request by Aladdin-ADD**
#7032

**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
updated a test case.

**Does this PR introduce a breaking change?**
not sure, I guess no. 😄 

**What needs to be documented once your changes are merged?**

### CODE REVIEW

File: lib/ProvidePlugin.js [hunk 1]
```diff
@@ -76,9 +76,11 @@ class ProvidePlugin {
 				normalModuleFactory.hooks.parser
 					.for("javascript/dynamic")
 					.tap("ProvidePlugin", handler);
-				normalModuleFactory.hooks.parser
-					.for("javascript/esm")
-					.tap("ProvidePlugin", handler);
+
+				// Disable ProvidePlugin for javascript/esm, see https://github.com/webpack/webpack/issues/7032
+				// normalModuleFactory.hooks.parser
```
Comment (sokra) [inline lib/ProvidePlugin.js]: remove the commented code, leave only a note why it's not enabled

**Comment by Aladdin-ADD** [comment]
@sokra  I updated the test case in `test/configCases/plugins/provide-plugin`, added a '.mjs' in it. 
please let me know if need something more.

File: test/configCases/plugins/provide-plugin/webpack.config.js [hunk 2]
```diff
@@ -10,7 +10,8 @@ module.exports = {
 			es2015_name: ["./harmony", "default"],
 			es2015_alias: ["./harmony", "alias"],
 			es2015_year: ["./harmony", "year"],
-			"this.aaa": "./aaa"
+			"this.aaa": "./aaa",
+			esm: "./fff.mjs"
```
Comment (sokra) [inline test/configCases/plugins/provide-plugin/webpack.config.js]: change this to `esm: "fail"`

import `fff.mjs` from `index.js`.

Use `esm` in `fff.mjs`. It should not be replaced. i. e.

``` js
export function() {
  return esm;
}
```

---

Currently you are not testing your change.

File: test/configCases/plugins/provide-plugin/index.js [hunk 3]
```diff
@@ -56,3 +56,8 @@ it("should provide ES2015 modules", function() {
 	expect((es2015_alias)).toBe("ECMAScript Harmony");
 	expect((es2015_year)).toBe(2015);
 });
+
+it("should not provide for mjs", function(){
+	var foo = require(__dirname + "/foo.mjs").default;
```
Comment (sokra) [inline test/configCases/plugins/provide-plugin/index.js]:
```suggestion
	var foo = require("./foo.mjs").default;
```

**Comment by sokra** [comment]
Thanks

**Pull request merged**